[Bryan Kwok Yan Fai] iP

kwokieee opened the pull request:
# Welcome to Duke!

> "Rule number one is, don't sweat the small stuff. Rule number two is, it's all small stuff." - Robert Eliot

Too many tasks to remember? ~~Time to cry in a corner~~ Duke is here for you!
Duke is your very own task manager, helping you remember whatever you need to do.

Use Duke in just **3 easy steps**!
1. Download the .jar file from [here](https://github.com/kwokieee/ip/releases/tag/v0.1)
2. Double-click to run it
3. Add your tasks! :relaxed:

Duke supports the following tasks types:
* ToDos
* Deadlines
* Events

Features:
- [x] Adding tasks
- [x] Completing tasks
- [x] Deleting tasks
- [x] Viewing current tasks
- [x] Finding tasks using keyword search
- [ ] GUI (coming soon!)

## Source Code
If you are familiar with Java, you could take a look at the [source code](https://github.com/kwokieee/ip) as well!
Duke is programmed using the *OOP paradigm*. Here's a snippet showing the `main` method in the `Duke` class:
```java
public class Duke {
    public static void main(String[] args) {
        new Duke().run();
    }
}
```

Commit:
Add Gradle support

kwokieee committed:
Complete Level 1. Greet, Echo, Exit

kwokieee committed:
Complete Level 2. Add, List

kwokieee committed:
Add Task.java

kwokieee committed:
Add mark as done functionality

kwokieee committed:
Add comments to Task.java

kwokieee committed:
Done command: add error checking

kwokieee committed:
Add ToDo.java

kwokieee committed:
Add Deadline.java

kwokieee committed:
Add Event.java

kwokieee committed:
Add todo, deadline and event commands

kwokieee committed:
Add test input

kwokieee committed:
Add expected result for test input

kwokieee committed:
Add error handling for todo, deadline and event

Users may not enter commands in the correct format.

To prevent the program from crashing, we display the proper format of a
command to the user should they enter an invalid format.

kwokieee committed:
Add DukeException.java

kwokieee committed:
Rewrite each command as a function

To provide better abstraction, tidiness of code and allow the usage of
DukeException, the code dealing with each command was abstracted out
into separate methods.

kwokieee committed:
Change taskList to ArrayList

kwokieee committed:
Add delete task functionality

kwokieee committed:
Add data directory

kwokieee committed:
Add alt constructor and getSaveFormat method

kwokieee committed:
DukeException: Change to extend RuntimeException

kwokieee committed:
Add save functionality

kwokieee committed:
Add test for save functionality

kwokieee committed:
Add check for delimiter in task description

kwokieee committed:
Add test for task description containing delimiter

kwokieee committed:
Modify date implementation for event and deadline

kwokieee committed:
Merge branch 'branch-Level-7'

kwokieee committed:
Merge branch 'branch-Level-8'

kwokieee committed:
Fix constructors for new date implementation

kwokieee committed:
Add tests for new date implementation

kwokieee committed:
Add Ui.java

kwokieee committed:
Add Storage.java

kwokieee committed:
Add TaskList.java

kwokieee committed:
Add enums for user commands

kwokieee committed:
Add Parser.java

kwokieee committed:
Add classes for various commands

kwokieee committed:
Refactor code to use more OOP

kwokieee committed:
Add check for marking a complete task as done

kwokieee committed:
Group classes into packages

kwokieee committed:
Edit to compile classes in all packages

kwokieee committed:
Add equals method for comparison in unit tests

kwokieee committed:
Fix error handling

Resolved the issue where parseSaveFile() did not handle the error of
the save file having an invalid format.

Removed an unnecessary check performed by parseUserInput() when
handling empty task descriptions.

kwokieee committed:
Add unit tests for Parser and Ui classes

kwokieee committed:
Create JAR

kwokieee committed:
Add equals method

kwokieee committed:
Add JavaDoc comments

kwokieee committed:
Fix to comply with Java coding standard

kwokieee committed:
Add find functionality

kwokieee committed:
Merge branch 'branch-A-JavaDoc'

kwokieee committed:
Merge branch 'branch-A-CodingStandard'

kwokieee committed:
Merge branch 'branch-Level-9'

kwokieee committed:
Fix JavaDoc comment whitespace

kwokieee committed:
Add JavaDoc comments for Duke.java

kwokieee committed:
Update to JUnit 5.7.0

kwokieee committed:
Merge branch 'add-gradle-support'

kwokieee committed:
Modify application mainClassName

kwokieee committed:
Add checkstyle config files

kwokieee committed:
Fix checkstyle violations

knotstoks reviewed (commented):
LGTM, I don't see anything wrong with the coding standards in the entire repo!

kwokieee committed:
Integrate JavaFX for Duke GUI

kwokieee committed:
Modify UiTest.java for JavaFX integration

kwokieee committed:
Fix checkstyle errors

kwokieee committed:
Merge branch 'branch-Level-10'

Jacobnbh reviewed (commented):
LGTM! Code is succinct and readability is really good (nothing nested too deep, expressions were easy to understand). Naming guidelines were also well adhered to (no misleading names, length of names were good). Really learnt a lot from reading your code as well! 😄 👍

kwokieee committed:
Add assertions

kwokieee committed:
Parser class: refactor parseUserInput method

The parseUserInput method was over 30 LOC.

This does not follow good code quality guidelines due to poor
readability.

Let's let parseUserInput handle creating the appropriate commands,
and let the execute method in each command handle the formatting and
validation of the user's input.

Dr-Octavius commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -1,3 +1,11 @@
+import duke.task.*;
+import duke.ui.Ui;
+
 public class AddCommand extends Command {
```
Maybe addCommand should be a method within Command rather than a class on it's own?

Dr-Octavius commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -1,3 +1,18 @@
+package duke.parser;
+
+import duke.command.AddCommand;
```
Similar to previous comments, maybe the various command types can be separate methods within the Command Class, which reduces the need for having to repeat the isExit() method?

kwokieee replied:
Thanks for the suggestion! Agree that having to repeat the isExit() method is redundant and that the various commands can be rewritten in a more succinct manner.

Dr-Octavius commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -42,4 +42,16 @@ public void execute(TaskList taskList, Ui ui, Storage storage) {
     }
+
+    @Override
+    public boolean equals(Object other) {
```
Maybe for the equals method could follow the if, else if and else equals format provided by Java?

kwokieee replied:
Thanks for the suggestion! Following this format would make the code neater and more standardised.

kwokieee committed:
Update tests

Dr-Octavius commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -19,18 +19,22 @@
-            case "E":
-                return new Event(taskDescription, isDone, taskInfo[3]);
-            default:
+            try {
```
Great use of try/catch here😊

Dr-Octavius reviewed (commented):
Overall unit tests was written well. Consistent and straight to the point. Maybe a bit too much OOP use for the Command Class. Otherwise, LGTM! Kee up the good work!😊

kwokieee committed:
Merge pull request #1 from kwokieee/branch-A-Assertions

Add assertions

kwokieee committed:
Merge branch 'master' into branch-A-Assertions

kwokieee committed:
Merge branch 'master' into branch-A-CodeQuality

kwokieee committed:
Merge pull request #2 from kwokieee/branch-A-CodeQuality

Improve code quality

kwokieee committed:
Merge branch 'master' into branch-A-Assertions

kwokieee committed:
Merge branch 'master' into branch-A-CodeQuality

kwokieee committed:
Add basic CI workflow

kwokieee committed:
Add help menu GUI element

kwokieee committed:
Add help command functionality

kwokieee committed:
Add help button on GUI

kwokieee committed:
Change dialog from Label to Text

kwokieee committed:
Add JavaDoc comments

kwokieee committed:
Add unit tests for help command

kwokieee committed:
Fix checkstyle violations

kwokieee committed:
Merge pull request #3 from kwokieee/branch-C-Help

Add help feature

kwokieee committed:
Merge branch 'master' into branch-C-Help

kwokieee committed:
Merge branch 'master' into branch-A-Assertions

kwokieee committed:
Merge branch 'master' into branch-A-CodeQuality

kwokieee committed:
Merge branch 'branch-A-Assertions'

kwokieee committed:
Merge branch 'branch-A-CodeQuality'

kwokieee committed:
Merge branch 'branch-C-Help'

kwokieee committed:
Change background colors

kwokieee committed:
Reduce profile picture size

kwokieee committed:
Modify spacing

kwokieee committed:
Add help message on help button press

kwokieee committed:
Add title to window

kwokieee committed:
Fix checkstyle and unit test errors

kwokieee committed:
Add user guide

kwokieee committed:
Fix images not showing

kwokieee committed:
Fix line break

kwokieee committed:
Add bold text

kwokieee committed:
Set theme jekyll-theme-cayman

kwokieee committed:
Fix emoji not showing

kwokieee committed:
Merge branch 'master' of https://github.com/kwokieee/ip

kwokieee committed:
Modify javafx dependencies

kwokieee committed:
Fix font error on mac
